File: os_with_rust/src/types/process_directory.rs
use uuid::Uuid;
use yew::virtual_dom::VNode;

/// reducer's Action
pub struct ProcessAction {
    pub process: ProcessState,
    pub action_type: String,
}

#[derive(Clone, PartialEq, Debug)]
pub struct Dimension {
    pub height: f64,
    pub width: f64,
    pub top: f64,
    pub left: f64,
}

/// reducer's State
#[derive(Clone, PartialEq, Debug)]
pub struct ProcessState {
    pub process_name: Option<String>,
    pub process: Option<VNode>,
    pub id: Option<Uuid>,
    pub dimension: Option<Dimension>,
    pub is_full_size: Option<bool>,
    pub temp_dimension: Option<Dimension>,
}

File: my-solid-project/src/types/processDirectory.ts
import { JSX } from "solid-js/jsx-runtime";

export type ThemeContextState = {
  readonly color: string;
  readonly title: string;
};

export type ProcessesContextValue = [
  state: ProcessState[],
  actions: {
    addProcess: (id: string) => void;
    deleteProcess: (id: string) => void;
    changeProcessDimension: (id: string, dimension: Dimension) => void;
    enlarge: (id: string) => void;
    changeActive: (id: string) => void;
    shrink: (id: string) => void;
    openFile: (processId: string, filePath: string) => void;
  }
];

export interface ProcessState {
  processName: string;
  process: JSX.Element;
  id: string; // UUID
  dimension: Dimension;
  isFullSize: boolean;
  tempDimension?: Dimension;
  iconPath: string;
  active: boolean;
  menus?: Menus[];
  isShrunk: boolean;
}

export interface Menus {
  starting: string;
  menus: Menu[];
}

export interface Menu {
  [key: string]: string | Menu;
}

export interface Dimension {
  left: number;
  top: number;
  width: number;
  height: number;
}
